src/borg/archive.py

```diff
@@ -468,6 +470,8 @@ def save(self, name=None, comment=None, timestamp=None, additional_metadata=None
         data = self.key.pack_and_authenticate_metadata(metadata.as_dict(), context=b'archive')
         self.id = self.key.id_hash(data)
         self.cache.add_chunk(self.id, Chunk(data), self.stats)
+        self.repository.async_wait()
```

Comment:
i think the only way to avoid too much stuff piling up at some place is to stop processing stuff if we notice we have too much in flight. and that should not be a perf problem as long as window_size * datasize > bandwidth * latency (or so). of course we currently do not know most of these variables....

for the special issue of async_wait waiting a long time to receive all the errors (in case everything is going wrong, but was delayed that much so it was not received before we got into async_wait):
async_wait could get a param fail_fast (default True) and would raise exceptions by itself if fail_fast is True.